airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py

```diff
@@ -27,6 +27,75 @@ def get_docker_hub_auth_token() -> str:
     return token
 
 
+def get_docker_hub_headers() -> Dict | None:
+    if "DOCKER_HUB_USERNAME" not in os.environ or "DOCKER_HUB_PASSWORD" not in os.environ:
+        # If the Docker Hub credentials are not provided, we can only anonymously call the Docker Hub API.
+        # This will only work for public images and lead to a lower rate limit.
+        return {}
+    else:
+        token = get_docker_hub_auth_token()
+        return {"Authorization": f"JWT {token}"} if token else {}
+
+
+def get_docker_hub_tags_and_digests(
+    image_name: str,
+    retries: int = 0,
+    wait_sec: int = 30,
+    next_page_url: str | None = None,
+    tags_and_digests: Dict[str, str] | None = None,
+    paginate: bool = True,
+) -> Dict[str, str]:
+    """Find all released tags and digests for an image.
+
+    Args:
+        image_name (str): The image name to get tags and digest
+        retries (int, optional): The number of times to retry the request. Defaults to 0.
+        wait_sec (int, optional): The number of seconds to wait between retries. Defaults to 30.
+        next_page_url (str | None, optional): The next DockerHub page to consume. Defaults to None.
+        tags_and_digest (Dict[str, str] | None, optional): The accumulated tags and digests for recursion. Defaults to None.
+
+    Returns:
+        Dict[str, str]: Mapping of image tag to digest
+    """
+    headers = get_docker_hub_headers()
+    tags_and_digests = tags_and_digests or {}
+
+    if not next_page_url:
+        tags_url = f"https://registry.hub.docker.com/v2/repositories/{image_name}/tags"
+    else:
+        tags_url = next_page_url
+
+    # Allow for retries as the DockerHub API is not always reliable with returning the latest publish.
+    for _ in range(retries + 1):
+        response = requests.get(tags_url, headers=headers)
+        if response.ok:
+            break
+            # This is to handle the case when a connector has not ever been released yet.
+        if response.status_code == 404:
+            return tags_and_digests
+        time.sleep(wait_sec)
+
+    json_response = response.json()
+    tags_and_digests.update({result["name"]: result.get("digest") for result in json_response.get("results", [])})
+    if paginate:
+        if next_page_url := json_response.get("next"):
+            tags_and_digests.update(
+                get_docker_hub_tags_and_digests(
+                    image_name, retries=retries, wait_sec=wait_sec, next_page_url=next_page_url, tags_and_digests=tags_and_digests
+                )
+            )
+    return tags_and_digests
+
+
+def get_latest_version_on_dockerhub(image_name: str) -> str | None:
+    tags_and_digests = get_docker_hub_tags_and_digests(image_name, retries=3, wait_sec=30)
+    if latest_digest := tags_and_digests.get("latest"):
```

Comment:
Not all of our docker repos have a `"latest"` tag, and we don't really rely on `"latest"` for getting the prod version of a connector (which is a good thing). So I feel like we shouldn't rely on it here, either. Can we check for the max version instead?

Reply:
All connector images published via `airbyte-ci` have `latest` tag:
https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py#L144

Reply:
> So I feel like we shouldn't rely on it here, either. Can we check for the max version instead?

As we also release pre-release image with tag which are not semver it's not possible to accurately determine the max version.
I think it's more correct to:
* Find the digest of the `latest` tag
* Find the version tag which match this digest and consider this version the latest one.

It more accurately matches what's happening in our publish pipeline.

Reply:
We can filter out pre-releases though, because they have a different naming convention.

Theoretically if `latest` was associated with v1.0 but there was a v1.1 in dockerhub, we'd want to increment past 1.1, right?

Reply:
Yes but that's not meant to happen as we always make `latest` match the latest version. If it happens it means we have a problem somewhere and this check might help catching it.

Reply:
I think if we want this check to catch a problem like that we should sort by tag and then verify that the digest of the max tag is the same as "latest".

Reply:
@clnoll that would be a reasonable check to add, but in POST_UPLOAD_VALIDATORS (running at the end of the publish pipeline), not in PRE_UPLOAD_VALIDATORS.

Reply:
@alafanechere I don't want to block you so I'll go ahead and approve, but I'd like to see us moving away from using or relying on "latest", as it's not best practice due to the ambiguity associated with it. I think this conversation actually demonstrates that in a way, but obviously our usage of it is a bigger problem than just this PR.